### PR TITLE
Implement automatic nonce prepending/extracting

### PR DESCRIPTION
**Pull request by btoews**
This PR implements automatic nonce prefixing and the `New([]byte) (cipher.AEAD, error)` constructor. I wen't with a wrapper struct for implementing the different behavior, but can switch to a `manual bool` field on the `xaes256gcm` struct if you'd rather.

### CODE REVIEW

File: XAES-256-GCM.go [hunk 1]
```diff
@@ -28,6 +29,66 @@ const OverheadWithManualNonces = 16
...
+	}
+
+	if total := len(dst) + len(plaintext) + Overhead; cap(dst) < total {
+		tmp := make([]byte, len(dst), total)
+		copy(tmp, dst)
+		dst = tmp
+	}
+
+	nonce = dst[len(dst) : len(dst)+NonceSize]
+	if _, err := rand.Read(nonce); err != nil {
```
Comment (FiloSottile) [inline XAES-256-GCM.go:67]: If the tmp path is not taken, and `dst = plaintext[:0]`, this will overwrite the plaintext.

File: XAES-256-GCM.go [hunk 2]
```diff
@@ -28,6 +29,66 @@ const OverheadWithManualNonces = 16
...
+		panic(err)
+	}
+
+	dst = dst[:len(dst)+NonceSize]
+
+	return x.xaes256gcmManual.Seal(dst, nonce, plaintext, additionalData)
```
Comment (FiloSottile) [inline XAES-256-GCM.go:73]: If the tmp path is not taken, and `dst = plaintext[:0]`, this will panic due to inexact overlap.

**Review by FiloSottile — CHANGES_REQUESTED** [review]
Unfortunately, the automatic nonce mode interacts poorly with the AEAD interface. I think the only way without changing the stdlib is always doing an allocation and a copy, regrettably. We also need a lot more tests for these edge cases.
